src/psbt/mod.rs

```diff
@@ -37,3 +37,66 @@ impl PsbtUtils for PSBT {
         }
     }
 }
+
+#[cfg(test)]
+mod test {
+    use crate::bitcoin::consensus::deserialize;
+    use crate::bitcoin::TxIn;
+    use crate::psbt::PSBT;
+    use crate::wallet::test::{get_funded_wallet, get_test_wpkh};
+    use crate::wallet::AddressIndex;
+    use crate::SignOptions;
+
+    // from bip 174
+    const PSBT_STR: &str = "cHNidP8BAKACAAAAAqsJSaCMWvfEm4IS9Bfi8Vqz9cM9zxU4IagTn4d6W3vkAAAAAAD+////qwlJoIxa98SbghL0F+LxWrP1wz3PFTghqBOfh3pbe+QBAAAAAP7///8CYDvqCwAAAAAZdqkUdopAu9dAy+gdmI5x3ipNXHE5ax2IrI4kAAAAAAAAGXapFG9GILVT+glechue4O/p+gOcykWXiKwAAAAAAAEHakcwRAIgR1lmF5fAGwNrJZKJSGhiGDR9iYZLcZ4ff89X0eURZYcCIFMJ6r9Wqk2Ikf/REf3xM286KdqGbX+EhtdVRs7tr5MZASEDXNxh/HupccC1AaZGoqg7ECy0OIEhfKaC3Ibi1z+ogpIAAQEgAOH1BQAAAAAXqRQ1RebjO4MsRwUPJNPuuTycA5SLx4cBBBYAFIXRNTfy4mVAWjTbr6nj3aAfuCMIAAAA";
+
+    #[test]
+    #[should_panic(expected = "InputIndexOutOfRange")]
+    fn test_psbt_malformed_psbt_input_legacy() {
+        let psbt_bip: PSBT = deserialize(&base64::decode(PSBT_STR).unwrap()).unwrap();
+        let (wallet, _, _) = get_funded_wallet(get_test_wpkh());
+        let send_to = wallet.get_address(AddressIndex::New).unwrap();
+        let mut builder = wallet.build_tx();
+        builder.add_recipient(send_to.script_pubkey(), 10_000);
+        let (mut psbt, _) = builder.finish().unwrap();
+        psbt.inputs.push(psbt_bip.inputs[0].clone());
+        let options = SignOptions {
+            trust_witness_utxo: true,
+            assume_height: None,
+        };
+        let _ = wallet.sign(&mut psbt, options).unwrap();
```

Comment:
Relying on this `unwrap` to indicate that the test passes while having `unwrap` calls further up the test can give a false positive i.e., the test can pass due to one of the earlier `unwrap`s panicking.

Using an assertion on `is_err` and removing the `should_panic` makes the test more robust.

```
assert!(wallet.sign(&mut psbt, options).is_err());
```
But looses the error type check, I came up with the rather ugly:
```
        match wallet.sign(&mut psbt, options) {
            Err(Error::Signer(e)) => assert_eq!(e, SignerError::InputIndexOutOfRange),
            Err(e) => panic!("Wrong error type: {:?}", e),
            Ok(_) => panic!("sign() should have failed"),
        }
```

Reply:
Yes, you are right, but it's a lot of code...

If we want to be more precise, I think we should consider using https://docs.rs/crate/assert_matches/1.5.0 as a dev-dep which will also be standardized one day  https://doc.rust-lang.org/nightly/std/macro.assert_matches.html

it would become something like:
`assert_matches!(wallet.sign(&mut psbt, options).unwrap_err(), SignerError::InputIndexOutOfRange)`

@afilini ?

Reply:
I think we have a lot of tests that use the `should_panic`, so I wouldn't bother doing anything differently here.

It makes sense to add `assert_matches`, but that should be a different PR that also updates all the other tests.

Also ideally we should avoid unwrapping completely if we use it, so the check could be against the `Result` as well instead of just the error variant:

`assert_matches!(wallet.sign(&mut psbt, options), Err(Error::Signer(SignerError::InputIndexOutOfRange)))`